src/blockCache/sqlite.ts

```diff
@@ -0,0 +1,124 @@
+import Sqlite from "better-sqlite3";
+import { BlockCache, Block } from "../blockCache.js";
+
+const defaultTableName = "blocks";
+
+export type Options =
+  | {
+      dbPath: string;
+      tableName?: string;
+    }
+  | { db: Sqlite.Database; tableName?: string };
+
+interface Row {
+  chainId: number;
+  blockNumber: string;
+  timestamp: number;
+}
+
+export function createSqliteBlockCache(opts: Options): BlockCache {
+  let db: Sqlite.Database | null = null;
+
+  if (opts.tableName != null && opts.tableName.match(/[^a-zA-Z0-9_]/)) {
+    throw new Error(
+      `Table name ${opts.tableName} contains invalid characters. Only alphanumeric and underscore characters are allowed.`
+    );
+  }
+
+  const tableName = opts.tableName ?? defaultTableName;
+
+  return {
+    async init() {
+      if (db) {
+        throw new Error("Already initialized");
+      }
+
+      if ("db" in opts) {
+        db = opts.db;
+      } else {
+        db = new Sqlite(opts.dbPath);
+      }
+
+      db.exec("PRAGMA journal_mode = WAL;");
+
+      // TODO: Add proper migrations, with Kysely?
+      db.exec(
+        `CREATE TABLE IF NOT EXISTS ${tableName} (
+          chainId INTEGER,
+          blockNumber TEXT,
+          timestamp INTEGER,
+          PRIMARY KEY (chainId, blockNumber)
+        )`
+      );
+
+      db.exec(
+        `CREATE INDEX IF NOT EXISTS idx_chainId_timestamp_blockNumber 
+         ON ${tableName} (chainId, timestamp, blockNumber DESC);`
+      );
+    },
+    async getBlockByNumber(chainId, blockNumber) {
+      if (!db) {
+        throw new Error("SQLite database not initialized");
+      }
+
+      const row = db
+        .prepare(
+          `SELECT * FROM ${tableName} WHERE chainId = ? AND blockNumber = ?`
+        )
+        .get(chainId, blockNumber.toString()) as Row | undefined;
```

Comment:
Casting might be minimized here and elsewhere and flow made explicit with something like

```ts
if (row === undefined) {
  return null
} else {
  return ... as Block
}